fix(components): add ref to Autocomplete component

gregorijobber opened the pull request:
<!--
  Atlantis uses Conventional Commits to track versions.
  Pull request titles should follow the following format.

  For help creating your pull request, you can [use this tool](https://atlantis.getjobber.com/pull-request-name-generator)

  <TYPE>(<optional SCOPE>): <conditionally BREAKING CHANGE:> <description>

  eg.
    fix(SCOPE): stop graphite breaking when too much pressure applied — Patch Release
    feat(SCOPE): add 'graphiteWidth' option — (Minor) Feature Release
    feat(SCOPE): BREAKING CHANGE: remove graphiteWidth option — (Major) Breaking Release

  TYPE should consist of:
    - fix: a commit of the type fix patches a bug in your codebase
    - feat: a commit of the type feat introduces a new feature to the codebase
    - docs: documentation only changes
    - build: improvements to the build system
    - refactor: a change that neither fixes a bug nor introduces a feature
    - chore: other changes that don't modify src or test files

  SCOPE should be one of:
    - components
    - components-native
    - design
    - eslint
    - generators
    - hooks
    - stylelint


  If your pull request introduces a breaking change please append `BREAKING CHANGE:` following type / scope.

  Further Reading:
    - https://www.conventionalcommits.org
    - https://github.com/commitizen/conventional-commit-types/blob/master/index.json
-->

## Motivations

We need that to be able to use `ref` from `Autocomplete` component, so we can use actions like `focus` and `scrollIntoView` to create interactivity with the input

## Changes

<!-- https://keepachangelog.com/en/1.0.0/ -->

### Added

- Added `ref` prop to parent component (`Autocomplete`) to be used as reference for actions

### Changed

- <!-- changes in existing functionality -->

### Deprecated

- <!-- soon-to-be removed features -->

### Removed

- <!-- now removed features -->

### Fixed

- <!-- for any bug fixes -->

### Security

- <!-- in case of vulnerabilities -->

## Testing

<!-- How to test your changes. -->

---

[In Atlantis we use Github's built in pull request reviews](https://help.github.com/en/articles/about-pull-request-reviews).

![Random photo of Atlantis](https://loremflickr.com/672/400/atlantis)

MichaelParadis commented:
@gregorijobber Because this is a fix for components the PR title should be `fix(components): ...`

mattander reviewed (commented):
I've tested it out and it works well. Do we want to add a test to check that ref is being passed up and has the expected methods on it?

MichaelParadis reviewed (changes requested):
Minor comments, had some environment issues but they look to be resolved so I will be continuing testing

darryltec commented on packages/components/src/Autocomplete/Autocomplete.tsx:
```diff
@@ -69,20 +69,23 @@ interface AutocompleteProps
-  validations,
-  ...inputProps
-}: AutocompleteProps) {
+function AutocompleteInternal(
```
hmmm, it should work. I believe @jessedoyle leverages that

MichaelParadis replied:
🤔 we should reach out and talk to him. I am surprised it worked because it "should" be a forward ref from my understanding of refs from the [docs](https://react.dev/reference/react/forwardRef#forwarding-a-ref-through-multiple-components)

gregorijobber replied:
In my tests it didn't work at anytime without adding the `forwardRef` as per @MichaelParadis suggestion!

mattander replied:
I'd be shocked if this works without `forwardRef` unless I've really misunderstood the docs.

darryltec replied:
Ever noticed how you don't even have to declare `ref` as a prop. It just happens the moment you `forwardRef`. It'll work without it, the downside is that you have to manually expose it

> Ref forwarding is a technique for automatically passing a [ref](https://legacy.reactjs.org/docs/refs-and-the-dom.html) through a component to one of its children. This is typically not necessary for most components in the application

[source](https://legacy.reactjs.org/docs/forwarding-refs.html#:~:text=Ref%20forwarding%20is%20a%20technique%20for%20automatically%20passing%20a%20ref%20through%20a%20component%20to%20one%20of%20its%20children.%20This%20is%20typically%20not%20necessary%20for%20most%20components%20in%20the%20application)

Now if they do change something in v19, that'll get fucked haha so we need to expose the element on the ref as well instead of a limited version of it.

mattander replied:
Ah okay, that does make sense. 

In terms of leaving this/removing it, maybe we should leave it for now but deprecate that approach and get folks to just use the forwarded ref?

MichaelParadis replied:
That is a good compromise @mattander

darryltec replied:
leave it as is, if this PR passes review then we merge it and we'll mark that other thing deprecated. Otherwise, if there are more changes, we could add that deprecation task here

gregorijobber replied:
Ok, now I'm confused 😆 should we remove the `inputRef` then? @MichaelParadis @darryltec are you ok with that? Then I can update the PR for your approval and we can move on 🥳

MichaelParadis replied:
Based on @jessedoyle 's comment I think deprecation and removing in the future makes sense in case others are using it

darryltec replied:
Yup! Going into more detail about it. Deprecation and removal isn't really a straightforward thing as we do need to replace that functionality if we are removing it.

gregorijobber commented:
@MichaelParadis I've addressed most of your comments. I've added tests for `focus` and `scrollIntoView` cases for `InputText` and `Autocomplete` components.

@darryltec should I remove or just leave that property `inputRef` then?

mattander reviewed (approved):
Okay, I've tested the latest and it still works. Good to go from my perspective.

MichaelParadis reviewed (approved):
LGTM! Going to wait on @darryltec before merging
